### PR TITLE
Sailfish OS  device build configuration connected with touch screen support required for device with this OS

### PR DESCRIPTION
**Pull request by wiktorek140**
## Description

This commits include spec file for building rpm for Sailfish OS .
In addidion added support for touchscreen in wayland input

## Reviewers

@hiddenasbestos @twinaphex

### CODE REVIEW

**Comment by twinaphex** [comment]
Thanks for this contribution!

One thing I am wondering though, will the mouse also still continue to work on PC Wayland based desktops? Or does only touch work now?

**Comment by wiktorek140** [comment]
It should because i create oter function for touch, so it shouldnt confict with mouse. but i dont have device with wayland desktop to test it

File: input/drivers/wayland_input.c [hunk 1]
```diff
@@ -98,6 +99,31 @@ static int16_t input_wl_lightgun_state(input_ctx_wayland_data_t *wl, unsigned id
    return 0;
 }
 
+//static void wayland_poll_touch(void *data)
```
Comment (bparker06) [inline input/drivers/wayland_input.c]: C89

File: gfx/drivers_context/wayland_ctx.c [hunk 2]
```diff
@@ -329,8 +343,134 @@ static const struct wl_pointer_listener pointer_listener = {
...
+}
+static void touch_handle_frame(void *data,
+      struct wl_touch *wl_touch)
+{
+   /* TODO */
+}
+static void touch_handle_cancel(void *data,
+      struct wl_touch *wl_touch)
+{
+   //If i understand the spec correctly we have to reset all touches here
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: C89

File: gfx/drivers_context/wayland_ctx.c [hunk 3]
```diff
@@ -329,8 +343,134 @@ static const struct wl_pointer_listener pointer_listener = {
...
+      wl_fixed_t x,
+      wl_fixed_t y)
+{
+
+   gfx_ctx_wayland_data_t *wl = (gfx_ctx_wayland_data_t*)data;
+   if (num_active_touches < MAX_TOUCHES)
+   {
+     for (int i=0;i<MAX_TOUCHES;i++)
+     {
+       // Use next empty slot
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: C89

File: gfx/drivers_context/wayland_ctx.c [hunk 4]
```diff
@@ -51,6 +51,19 @@
...
+
+typedef struct touch_pos
+{
+    bool active;
+    int32_t id;
+    unsigned x;
+    unsigned y;
+} touch_pos_t;
+
+int num_active_touches;
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: don't make this global to the entire app, use static

File: gfx/drivers_context/wayland_ctx.c [hunk 5]
```diff
@@ -51,6 +51,19 @@
...
+touch_pos_t active_touch_positions[MAX_TOUCHES];
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: use static

**Comment by twinaphex** [comment]
I can test it out later today on Linux with Wayland/Weston.

**Comment by wiktorek140** [comment]
@twinaphex i installed wayland/weston nad tested it. mouse work. u can click and other things but mousce scroll dont work. i will test that, if work previously i will look for workaround for touch/mouse

**Comment by wiktorek140** [comment]
@twinaphex if my device is configured correctly, mouse scroll dont work too on clean build. of course in menu. i cannot tell more about mouse in games

File: gfx/common/wayland_common.h [hunk 6]
```diff
@@ -28,6 +28,16 @@
 #define UDEV_KEY_MAX			0x2ff
 #define UDEV_MAX_KEYS (UDEV_KEY_MAX + 7) / 8
 
+#define MAX_TOUCHES             16
+
+typedef struct
+{
+  bool active;
```
Comment (bparker06) [inline gfx/common/wayland_common.h]: need 3 spaces

File: gfx/drivers_context/wayland_ctx.c [hunk 7]
```diff
@@ -329,8 +343,136 @@ static const struct wl_pointer_listener pointer_listener = {
...
+static void touch_handle_down(void *data,
+      struct wl_touch *wl_touch,
+      uint32_t serial,
+      uint32_t time,
+      struct wl_surface *surface,
+      int32_t id,
+      wl_fixed_t x,
+      wl_fixed_t y)
+{
+   int i;
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: need 3 spaces

File: gfx/drivers_context/wayland_ctx.c [hunk 8]
```diff
@@ -329,8 +343,136 @@ static const struct wl_pointer_listener pointer_listener = {
...
+   gfx_ctx_wayland_data_t *wl = (gfx_ctx_wayland_data_t*)data;
+   if (num_active_touches < MAX_TOUCHES)
+   {
+     for (i=0;i<MAX_TOUCHES;i++)
+     {
+       // Use next empty slot
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: C89

File: gfx/drivers_context/wayland_ctx.c [hunk 9]
```diff
@@ -329,8 +343,136 @@ static const struct wl_pointer_listener pointer_listener = {
...
+         active_touch_positions[num_active_touches].y = (unsigned) wl_fixed_to_int(y);
+         num_active_touches++;
+         break;
+       }
+     }
+   }
+}
+static void reorder_touches()
+{
+   int i,j;
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: need 3 spaces

File: gfx/drivers_context/wayland_ctx.c [hunk 10]
```diff
@@ -329,8 +343,136 @@ static const struct wl_pointer_listener pointer_listener = {
...
+      }
+   }
+}
+static void touch_handle_up(void *data,
+      struct wl_touch *wl_touch,
+      uint32_t serial,
+      uint32_t time,
+      int32_t id)
+{
+   int i;
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: need 3 spaces

File: gfx/drivers_context/wayland_ctx.c [hunk 11]
```diff
@@ -329,8 +343,136 @@ static const struct wl_pointer_listener pointer_listener = {
...
+   reorder_touches(wl);
+}
+static void touch_handle_motion(void *data,
+      struct wl_touch *wl_touch,
+      uint32_t time,
+      int32_t id,
+      wl_fixed_t x,
+      wl_fixed_t y)
+{
+   int i;
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: need 3 spaces

File: gfx/drivers_context/wayland_ctx.c [hunk 12]
```diff
@@ -354,6 +496,17 @@ struct wl_seat *seat, unsigned caps)
       wl_pointer_destroy(wl->wl_pointer);
       wl->wl_pointer = NULL;
    }
+   if ((caps & WL_SEAT_CAPABILITY_TOUCH) && !wl->wl_touch)
+   {
+      wl->wl_touch = wl_seat_get_touch(seat);
```
Comment (bparker06) [inline gfx/drivers_context/wayland_ctx.c]: need 3 spaces

File: input/drivers/wayland_input.c [hunk 13]
```diff
@@ -98,6 +99,31 @@ static int16_t input_wl_lightgun_state(input_ctx_wayland_data_t *wl, unsigned id
...
+static void input_wl_touch_pool(void *data)
+{
+   int id;
```
Comment (bparker06) [inline input/drivers/wayland_input.c]: need 3 spaces

File: input/drivers/wayland_input.c [hunk 14]
```diff
@@ -98,6 +99,31 @@ static int16_t input_wl_lightgun_state(input_ctx_wayland_data_t *wl, unsigned id
...
+   input_ctx_wayland_data_t *wl = (input_ctx_wayland_data_t*)data;
+   if (!wl)
+      return;
+
+   unsigned touch_x;
```
Comment (bparker06) [inline input/drivers/wayland_input.c]: C89

File: input/drivers/wayland_input.c [hunk 15]
```diff
@@ -98,6 +99,31 @@ static int16_t input_wl_lightgun_state(input_ctx_wayland_data_t *wl, unsigned id
...
+   unsigned touch_y;
+   for (id=0;id<MAX_TOUCHES;id++)
+   {
+     if(wayland_context_gettouchpos(wl, id, &touch_x, &touch_y))
```
Comment (bparker06) [inline input/drivers/wayland_input.c]: need space after `if`, and 3 spaces indentation

File: input/drivers/wayland_input.c [hunk 16]
```diff
@@ -203,6 +232,59 @@ static int16_t input_wl_pointer_state(input_ctx_wayland_data_t *wl,
    return 0;
 }
 
+static int16_t input_wl_touch_state(input_ctx_wayland_data_t *wl,
+      unsigned idx, unsigned id, bool screen)
+{
+   struct video_viewport vp;
```
Comment (bparker06) [inline input/drivers/wayland_input.c]: need 3 spaces

**Comment by bparker06** [comment]
Hi @wiktorek140, there are still comments I made previously that were not addressed, and I added some more now also. Thanks

File: gfx/drivers_context/wayland_ctx.c (hunk 12 above)
Comment (wiktorek140) [reply gfx/drivers_context/wayland_ctx.c]: there too

**Comment by twinaphex** [comment]
Alright, I tested it on Linux with Weston and the mouse still seems to work fine.

**Comment by twinaphex** [comment]
So if those last few nits pointed out by @bparker06 can be fixed, then we can merge this.

**Comment by twinaphex** [comment]
Alrigh, I think things are more or less sufficient here, so no need to be postponing this any longer.

@wiktorek140 Thanks a lot for this contribution. Now merged!

**Comment by twinaphex** [comment]
@wiktorek140 Could you write a guide on how to build this for sailfish OS? We could then place it on docs.libretro.com.

**Comment by wiktorek140** [comment]
Before start in your system you have to instaled virtualbox.

1. Install Sailfish OS SDK using this tutorial  - https://sailfishos.org/wiki/Application_SDK_Installation
2. Launch Saifish OS SDK, create sample project, and click hammer icon (to build it - easiest way to launch virtual machine with build environment) - now we have launched virtual machine.

3. Move to directory with your project (this one which you sed during insttalation)
4. Clone/copy/extract Retroarch source file (tobtained from https://github.com/libretro/RetroArch)
to this directory 
Warning! It don't work in any other place

5. If under linux: get into terminal and execute this command: 
"ssh -p 2222 -i ~/SailfishOS/vmshare/ssh/private_keys/engine/mersdk mersdk@localhost"

where ~ you can replace with path to SailfishSDK installation if you dont use default
6. in build environment move to this directory: "/home/src1/<Retroarch>" 
where <Retroarch> is extracted directory with source code

7. then execute: "mb2 -t SailfishOS-<2.2.0.29>-armv7hl -s pkg/sailfishos/retroarch-sailfishos.spec build"

<2.2.0.29> - replace it if required with version provided when ssh int virtual machine. its look like:
,---
| Sailfish OS *2.2.0.29* (Mouhijoki)
'---

8. It may ask to install additinal packages, simply click Y
9. If dependencies are okay then it will be builded in some time.
10. ready to install .rpm is located in RPMS/, its accesible via virtual machine or local machine, and it's ready to sideload/copy and install on device.

**Pull request merged**